compiler/rustc_lint/src/types.rs

```diff
@@ -136,6 +137,37 @@ declare_lint! {
     "detects invalid floating point NaN comparisons"
 }
 
+declare_lint! {
+    /// The `invalid_pointer_trait_comparisons` lint checks comparison
+    /// of `*const dyn Trait` as the operands.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # struct A;
+    /// # struct B;
+    ///
+    /// # trait T {}
+    /// # impl T for A {}
+    /// # impl T for B {}
+    ///
+    /// let ab = (A, B);
+    /// let a = &ab.0 as *const dyn T;
+    /// let b = &ab.1 as *const dyn T;
+    ///
+    /// let _ = a == b;
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// The comparison include metadata which is not guaranteed to be equivalent.
+    INVALID_POINTER_TRAIT_COMPARISONS,
+    Warn,
+    "detects invalid pointer trait comparisons"
```

Comment:
"invalid" is a strong word to use here. The comparisons aren't really invalid, are they? They just behave in surprising ways.